lib/ProductOpener/Images.pm

```diff
@@ -467,32 +483,51 @@ sub process_image_upload($$$$$$) {
 			my $filename = get_fileid(remote_addr(). '_' . $`);
 
 			my $current_product_ref = retrieve_product($code);
-			$imgid = $current_product_ref->{max_imgid} + 1;
 
-			# if for some reason the images directories were not created at product creation (it can happen if the images directory's permission / ownership are incorrect at some point)
-			# create them
+			# if for some reason the images directories were not created
+			# at product creation (it can happen if the images directory's
+			# permission / ownership are incorrect at some point),
+			# create the the directories for the product
+			make_path("$image_base_path/", { chmod => 0755 });
 
-			# Create the directories for the product
-			foreach my $current_dir  ($www_root . "/images/products") {
-				(-e "$current_dir") or mkdir($current_dir, 0755);
-				foreach my $component (split("/", $path)) {
-					$current_dir .= "/$component";
-					(-e "$current_dir") or mkdir($current_dir, 0755);
-				}
+			my $hash = Digest::SHA3->new(224)->addfile($file, 'b')->b64digest;
+			$imgid = time() . "." . $hash;
+			local $log->context->{imgid} = $imgid;
+			my $img_path = "$image_base_path/$imgid.$extension";
+			# Check that we don't already have the image with the same name
+			if (glob("$image_base_path/*.$hash.$extension")) {
+				return -3;
 			}
 
-			my $lock_path = "$www_root/images/products/$path/$imgid.lock";
-			while (-e $lock_path) {
-				$imgid++;
-				$lock_path = "$www_root/images/products/$path/$imgid.lock";
+			# Check that there's no other file with the same hash
+			$log->debug("comparing existing images with hash of new image", { path => $img_path }) if $log->is_debug();
+			my @files = glob("$image_base_path/*.*");
+			foreach my $existing_image_path (@files) {
+				next if not -f $existing_image_path;
+
+				my $existing_image_hash = Digest::SHA3->new(224)->addfile($existing_image_path, 'b')->b64digest;
+				$log->debug("comparing image", { existing_image_path => $existing_image_path, existing_image_hash => $existing_image_hash }) if $log->is_debug();
+				if ($existing_image_hash eq $imgid) {
+					$log->debug("image with same hash detected", { existing_image_path => $existing_image_path, existing_image_hash => $existing_image_hash }) if $log->is_debug();
+					# check the image was stored inside the
+					# product, it is sometimes missing
+					# (e.g. during crashes)
+					my $product_ref = retrieve_product($code);
+					if ((defined $product_ref) and (defined $product_ref->{images}) and (exists $product_ref->{images}{$imgid})) {
+						$$imgid_ref = $imgid;
+						return -3;
+					}
+					else {
+						$log->warn("image missing in product.sto");
+					}
+				}
 			}
 
-			local $log->context->{imgid} = $imgid;
 			$log->debug("new imgid determined") if $log->is_debug();
 
+			my $lock_path = "$image_base_path/$imgid.lock";
```

Comment:
The aim of the lock was to prevent 2 images with the same imgid to be created (when an user sends multiple images and 2 different perl processes run product_image_upload.pl, if we use a hash, we don't need the lock.